com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceCreateOrUpdateOperationsExecution.java

```diff
@@ -1,92 +1,128 @@
+            return;
+        }
+
+        if (lastServiceOperation.getState() == ServiceOperationState.FAILED) {
+            handleFailedState(execution.getStepLogger(), service, lastServiceOperation);
+        }
+    }
+
+    private String getSuccessMessage(CloudServiceExtended service, ServiceOperationType type) {
+        switch (type) {
+            case CREATE:
+                return MessageFormat.format(Messages.SERVICE_CREATED, service.getName());
+            case UPDATE:
+                return MessageFormat.format(Messages.SERVICE_UPDATED, service.getName());
+            default:
+                throw new IllegalStateException(MessageFormat.format(com.sap.cloud.lm.sl.cf.core.message.Messages.ILLEGAL_SERVICE_OPERATION_TYPE,
+                                                                     type));
         }
-        return lastOperation;
     }
 
-    private ServiceOperation parseServiceOperationFromMap(Map<String, Object> serviceOperation) {
-        ServiceOperationType type = ServiceOperationType.fromString((String) serviceOperation.get(SERVICE_OPERATION_TYPE));
-        ServiceOperationState state = ServiceOperationState.fromString((String) serviceOperation.get(SERVICE_OPERATION_STATE));
-        String description = (String) serviceOperation.get(SERVICE_OPERATION_DESCRIPTION);
-        if (description == null && state == ServiceOperationState.FAILED) {
-            description = Messages.DEFAULT_FAILED_OPERATION_DESCRIPTION;
+    private void handleFailedState(StepLogger stepLogger, CloudServiceExtended service, ServiceOperation lastServiceOperation) {
+        if (shouldFail(service)) {
+            throw new SLException(getFailureMessage(service, lastServiceOperation));
         }
-        return new ServiceOperation(type, description, state);
+        stepLogger.warn(getWarningMessage(service, lastServiceOperation));
     }
 
+    private boolean shouldFail(CloudServiceExtended service) {
+        return !service.isOptional();
+    }
+
+    private String getFailureMessage(CloudServiceExtended service, ServiceOperation operation) {
+        switch (operation.getType()) {
+            case CREATE:
+                return MessageFormat.format(Messages.ERROR_CREATING_SERVICE, service.getName(), service.getLabel(), service.getPlan(),
+                                            operation.getDescription());
+            case UPDATE:
+                return MessageFormat.format(Messages.ERROR_UPDATING_SERVICE, service.getName(), service.getLabel(), service.getPlan(),
+                                            operation.getDescription());
+            default:
+                throw new IllegalStateException(MessageFormat.format(com.sap.cloud.lm.sl.cf.core.message.Messages.ILLEGAL_SERVICE_OPERATION_TYPE,
+                                                                     operation.getType()));
+        }
+    }
+
+    private String getWarningMessage(CloudServiceExtended service, ServiceOperation operation) {
```

Comment:
This method and the one above are exactly the same with the exception of the messages to be formatted.
Can the code be extracted and reused to remove duplication?